<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Heterogeneity Scoring on scRNA Sequencing Data Based on Missingness • OAR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<script src="lightswitch.js"></script><script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Heterogeneity Scoring on scRNA Sequencing Data Based on Missingness">
<meta name="description" content="OAR (observed at random) score is a measure of heterogeneity among cells, allowing cell prioritization for downstream applications. For best results, test a group of similar cells where you expect some heterogeneity - i.e. one cell type across various biological samples or conditions.">
<meta property="og:description" content="OAR (observed at random) score is a measure of heterogeneity among cells, allowing cell prioritization for downstream applications. For best results, test a group of similar cells where you expect some heterogeneity - i.e. one cell type across various biological samples or conditions.">
<meta property="og:image" content="https://davidsanin.github.io/OARscRNA/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">OAR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/introductory_vignette.html">Introductory Vignette for OARscRNA</a></li>
    <li><a class="dropdown-item" href="articles/detailed_tutorial.html">Step-by-step guide to running OAR</a></li>
    <li><a class="dropdown-item" href="articles/Gene_expression.html">Gene Expression Analysis based on OAR scores</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/davidsanin/OARscRNA/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header">
<img src="logo.png" class="logo" alt=""><h1 id="oarscrna-">OARscRNA <a class="anchor" aria-label="anchor" href="#oarscrna-"></a>
</h1>
</div>
<div class="section level2">
<h2 id="heterogeneity-scoring-on-scrnaseq-data-based-on-missingness">Heterogeneity scoring on scRNAseq data based on missingness<a class="anchor" aria-label="anchor" href="#heterogeneity-scoring-on-scrnaseq-data-based-on-missingness"></a>
</h2>
<p>The OAR (observed at random) score reveals cellular heterogeneity, allowing <strong>cell prioritization</strong> for downstream applications. The results of this test will highlight distinct cells within a dataset (cells with a high OAR score) for further exploration. For best results, test a group of similar cells where you expect some heterogeneity - <em>i.e. one cell type across various biological samples or conditions</em>. OAR scores are cluster agnostic (no cluster labels are required) and are robust across:</p>
<ul>
<li>Technologies</li>
<li>Technical batches/library preparations</li>
<li>Organisms</li>
<li>Sampling</li>
</ul>
<p>A <strong>positive OAR score</strong> within a dataset examined, highlights a cell with gene expression that appears more homogenous than other cells in the mixture and is consequently highly distinct cell. <strong>OAR score is a measure of heterogeneity among cells</strong>.</p>
</div>
<div class="section level2">
<h2 id="motivation">Motivation<a class="anchor" aria-label="anchor" href="#motivation"></a>
</h2>
<p>scRNAseq data is very sparse (50-90% of expression values are 0). Sparsity is <em>generally</em> attributed to technical limitations associated with capturing RNA molecules from individual cells. <strong>Some 0s are expected</strong>, and are a consequence of the Gamma-Poisson distribution of count data<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;Svensson, V. (2020). Droplet scRNA-seq is not zero-inflated &lt;a href="https://www.nature.com/articles/s41587-019-0379-5" class="external-link"&gt;DOI: 10.1038/s41587-019-0379-5&lt;/a&gt;&lt;/p&gt;'><sup>1</sup></a>, whereas “Drop-out” (when 0 occurs where a positve count is expected) is a problem associated with specific technologies (UMI- vs. nonUMI-based)<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;Cao, Y. &lt;em&gt;et. al.&lt;/em&gt; (2021). UMI or not UMI, that is the question for scRNA-seq zero-inflation &lt;a href="https://www.nature.com/articles/s41587-020-00810-6" class="external-link"&gt;DOI: 10.1038/s41587-020-00810-6&lt;/a&gt;&lt;/p&gt;'><sup>2</sup></a>.</p>
<p>Sparsity has been used to:</p>
<ul>
<li><a href="https://doi.org/10.1038/s41467-020-14976-9" class="external-link">Cluster cells</a></li>
<li><a href="https://academic.oup.com/bioinformatics/article/35/16/2865/5258099" class="external-link">Define variable genes</a></li>
</ul>
<p><strong>Is there something else we can learn about cellular identity from sparsity in scRNAseq data?</strong> <strong>Is sparsity (~missingness) occurring completely at random?</strong></p>
</div>
<div class="section level2">
<h2 id="test-overview">Test overview<a class="anchor" aria-label="anchor" href="#test-overview"></a>
</h2>
<p>At the core of the OAR score is the identification of missing data patterns, followed by comparing the distribution of genes expressed in the identified patterns in each cell individually.</p>
<div class="float">
<img src="reference/figures/Test_OV.png" alt="Base Test"><div class="figcaption">Base Test</div>
</div>
<p>To calculate the OAR score we:</p>
<ol style="list-style-type: decimal">
<li>Estimate <a href="https://en.wikipedia.org/wiki/Hamming_distance" class="external-link">Hamming distances</a> between binarized vectors of gene expression.</li>
<li>Group genes across missing data patterns defined as those with small (0-0.05) Hamming distance. Genes with unique patterns <em>- i.e. with no “neighbors”</em>, are grouped together.</li>
<li>Compare the distribution of gene expression across identified patterns for <em>each cell</em> with a <a href="https://en.wikipedia.org/wiki/Kruskal%E2%80%93Wallis_test" class="external-link">Kruskal-Wallis</a> test.</li>
<li>Scale the resulting <em>corrected p value</em> distributions <em>across all cells</em> to obtain the <strong>OAR score</strong>.</li>
</ol>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>To install the latest version of our package, run:</p>
<p><code>devtools::install_github("davidsanin/OARscRNA")</code></p>
<p>If you want to install our vignettes (takes a few minutes!), try:</p>
<p><code>devtools::install_github("davidsanin/OARscRNA", build_vignettes = TRUE)</code></p>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<p>To calculate an OAR score from a Seurat with <strong>default</strong> parameters run:</p>
<p><code>oar(data = seurat.obj)</code></p>
<p>Or from a matrix of unnormalized read counts, run:</p>
<p><code>oar(data = read.counts, seurat_v5 = F)</code></p>
<p>This will automatically filter low expression genes, identify a suitable tolerance for the hamming distance, and return the OAR score, corrected and uncorrected p values and percentage of missing data for each cell (column) in the supplied object.</p>
<ul>
<li>If a Seurat object is supplied, the results are added as columns in the <code>meta.data</code> slot.</li>
</ul>
<p>For full details on all parameters, including a step by step breakdown of the process, please visit our <a href="https://davidsanin.github.io/OARscRNA/">documentation</a> or view our vignettes with <code>browseVignettes(package = "OAR")</code>.</p>
</div>
<div class="section level2">
<h2 id="tutorials-and-applications">Tutorials and Applications<a class="anchor" aria-label="anchor" href="#tutorials-and-applications"></a>
</h2>
<div class="section level3">
<h3 id="quick-overview">Quick overview<a class="anchor" aria-label="anchor" href="#quick-overview"></a>
</h3>
<p>Follow our quick guide on running the analysis with a single line: <code><a href="articles/introductory_vignette.html">vignette("introductory_vignette")</a></code></p>
</div>
<div class="section level3">
<h3 id="cell-prioritization-for-downstream-analysis">Cell prioritization for downstream analysis<a class="anchor" aria-label="anchor" href="#cell-prioritization-for-downstream-analysis"></a>
</h3>
<p>Follow our step-by-step tutorial on exploring how we can identify highly activated plasmacytoid dendritic cells based on a high OAR score: <code><a href="articles/detailed_tutorial.html">vignette("detailed_tutorial")</a></code></p>
</div>
<div class="section level3">
<h3 id="model-gene-expression-data-at-the-single-cell-level">Model gene expression data at the single cell level<a class="anchor" aria-label="anchor" href="#model-gene-expression-data-at-the-single-cell-level"></a>
</h3>
<p>Identify genes responsible for high OAR scoring cells at a single cell resolution: <code><a href="articles/Gene_expression.html">vignette("Gene_expression")</a></code></p>
</div>
<div class="section level3">
<h3 id="remove-sources-of-variation-from-scrnaseq-data-analysis">Remove sources of variation from scRNAseq data analysis<a class="anchor" aria-label="anchor" href="#remove-sources-of-variation-from-scrnaseq-data-analysis"></a>
</h3>
<p>Regress out OAR scores calculated on raw counts and found to be associated with unwanted sources of variation (batches, low quality cells, individual samples) to remove confounding factors.</p>
</div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<p><em>scRNAseq implementation:</em> Chen, R., Moore, H., Gueguen, PM., Kelly, B., Fertig, EJ., Sanin, DE., (2025). Scoring Cellular Heterogeneity by interrogating Missingness in single-cell RNA-seq Data. <em>In press</em></p>
<p><em>Statistical proof:</em> Chen, R., Chung, YC., Basu, S., Shi, Q., (2024). Diagnostic Test for Realized Missingness in Mixed-type Data. Sankhya B, 86(1), 109-138. DOI: 10.1007/s13571-023-00317-5</p>
</div>
</div>


  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/davidsanin/OARscRNA/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/davidsanin/OARscRNA/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>GPL (&gt;= 3)</small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing OAR</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>David Sanin <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-0188-7267" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Ruizhe Chen <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-3924-3328" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Hannah Moore <br><small class="roles"> Author </small> <a href="https://orcid.org/0009-0006-8245-5368" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Dmitrijs Lvovs <br><small class="roles"> Author </small> <a href="https://orcid.org/0009-0003-2152-6853" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Paul Gueguen <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-2930-6073" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by David Sanin, Ruizhe Chen, Hannah Moore, Dmitrijs Lvovs, Paul Gueguen.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
