---
title: "Gene Expression Analysis based on OAR scores"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Gene Expression Analysis based on OAR scores}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = F}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
library(knitr)

hook_output <- knitr::knit_hooks$get("output")
knit_hooks$set(output = function(x, options) {
   lines <- options$output.lines
   if (is.null(lines)) {
     return(hook_output(x, options))  # pass to default hook
   }
   x <- unlist(strsplit(x, "\n"))
   more <- "..."
   if (length(lines)==1) {        # first n lines
     if (length(x) > lines) {
       # truncate the output, but add ....
       x <- c(head(x, lines), more)
     }
   } else {
     x <- c(more, x[lines], more)
   }
   # paste these lines together
   x <- paste(c(x, ""), collapse = "\n")
   hook_output(x, options)
 })

opts_chunk$set(eval=TRUE,
   ## text results
   echo=TRUE,
   results=c('markup', 'asis', 'hold', 'hide')[1],
   ## plots
   fig.path=c('figure', 'figure/minimal-')[1],
   fig.keep=c('high', 'none', 'all', 'first', 'last')[1],
   fig.align=c('center', 'left', 'right', 'default')[1],
   fig.show=c('hold', 'asis', 'animate', 'hide')[1],
   dev=c('pdf', 'png', 'tikz')[2],
   fig.width=5, fig.height=5, #inches
   fig.env=c('figure', 'marginfigure')[1],
   fig.pos=c('', 'h', 't', 'b', 'p', 'H')[1]
   )
```

```{r setup}
library(OAR)
```

```{r include=F, echo=F}
#path_to_file <- system.file("extdata", "pdcs.rds", package="OAR")
path_to_file <- "/Users/repository/Git/OARscore/inst/extdata/pdcs.rds"
```

# Overview

In this vignette we will explore what genes are associated with increased OAR scores. We will use our toy plasmacytoid dendritic cell dataset that you can download from our [github](https://github.com/davidsanin/OARscRNA/tree/main/inst/extdata).

## 1. Calculating OAR scores

We can run our calculation with a single line of code. We recommend you follow out introductory tutorials for a full description of this function. The result is the Seurat object with the `OARscore`, `KW.pvalue`, `KW.BH.pvalue` and `pct.missing` values in the `meta.data` slot.

```{r oar wrapper, output.lines=5}
library(Seurat)
sc.data <- oar(data = readRDS(file = path_to_file), 
               seurat_v5 = T, count.filter = 1,
               blacklisted.genes = NULL, suffix = "",
               mismatch = T, tolerance = 0.04,
               cores = parallelly::availableCores()-2)
```

To explore which cells/clusters have the most heterogeneity, we can visualize the results as in a UMAP projection.

```{r featureplot score deg, out.width="50%", fig.cap="Feature plot of OAR score"}
library(Seurat)
p1 <- FeaturePlot(
  sc.data, features = "OARscore", order = T, pt.size = 0.5, 
  min.cutoff = "q40", max.cutoff = "q90")
p1
```

Notice that cells with a High OAR score in this dataset cluster together.

## 2. Identify genes associated with OAR scores

In a typical scRNAseq analysis workflow, differential gene expression is calculated based on averaged counts across identified clusters. Although highly informative, this approach fundamentally looses the single cell resolution achieved through this technology. As an alternative to using cluster labels to model gene expression, which invites biases associated with identifying clusters, we propose to model gene expression based exclusively on the value of the OAR score. We implement this approach using glmGamPoi[^1], which has superior performance to other differential gene expression methods in terms of both speed and scaling factor calculation.

[^1]: glmGapPoi: Fit Gamma-Poisson Generalized Linear Models Reliably. [git](https://github.com/const-ae/glmGamPoi)

```{r Calculate gene expression}
oar_deg <- oar_deg(
    data = sc.data, seurat_v5 = T, score.name = "OARscore", 
    score = NULL, count.filter = 1,
    blacklisted.genes = NULL,
    splines = TRUE, degrees.freedom = 5,
    auto.threshold = TRUE, custom.tr = NULL)
  
```

A brief explanation of these parameters is presented here:

-   `data` may be an raw matrix of counts or a fully processed seurat object with an OAR score calculated, as indicated by the `seurat_v5` parameter. The name of the score should be indicated in the `score.name` parameter.
-   `score` only needs to be provided if starting from a count matrix and should be a vector of the calculated OAR scores in the same order as the columns of the count matrix.
-   `count.filter` controls the minimum percentage of cells expressing any given gene for that gene to be retained in the analysis. Defaults to 1.
-   `blacklisted.genes`, Allows you to provide a vector of uninformative gene names (i.e. non-coding genes) to exclude from the analysis. **This analysis automatically excludes all ribosomal genes**. Defaults to `NULL`.
-   `splines` and `degrees.freedom`, control whether splines should be fit across the OAR score (and how many, as set by `degrees.freedom`) to capture non-linear associations between OAR and gene expression. We recommend setting `splines` to `TRUE` and `degrees.freedom` between 3-5. If `splines` is set to `FALSE`, only monotonic associations will be retrieved.
-   `auto.threshold` and `custom.tr` control the false discovery rate (**FDR**) threshold to consider a gene significantly associated with the OAR score. `auto.threshold` set to `TRUE` (**recommended**) estimates this threshold based on the number of cells tested. Typically, given the high number of observations, p values calculated based on individual cells tend to be highly significant, even where little association is observed between gene expression and the predictive variable. If `auto.threshold` set to `FALSE`, then a custom p value threshold should be provided with `custom.tr`. Setting this value to typical ranges (i.e. 0.05-0.01) will likely result in spurious associations.

The results of the anslysis is a matrix with the regulated and filtered genes:

```{r result overview, output.lines=10}
oar_deg
```

For a full description of each column, visit the `glmGamPoi` documentation. For our purposes, the `name` and `adj_pval` columns (corresponding to genes and FDR, respectively) are the most important results.

## 3. Visualize results

Overall, this analysis identifies 135 genes significantly associated with OAR scores. If we plot the top 9 most significant genes alongside our previously calculated OAR scores, we can observe that we identified several candidates for further analysis:

```{r featureplot comparison, out.width="100%", fig.cap="Genes associated with OAR score"}
library(Seurat)
library(patchwork)
library(ggplot2)

p1 <- FeaturePlot(
  sc.data, features = "OARscore", order = T, pt.size = 1, 
  min.cutoff = "q40", max.cutoff = "q90") + 
  labs(title = "OAR score") +
  theme(
      aspect.ratio = 1,
      legend.position = "none",
      axis.title = element_blank(),
      axis.text = element_blank(),
      axis.ticks = element_blank(),
      axis.line = element_blank(),
      plot.title = element_text(size = 8, face = "plain"))

p2 <- FeaturePlot(
  object = sc.data, pt.size = 0.5,
  slot = "counts",
  min.cutoff = "q40", max.cutoff = "q90",
  features = oar_deg$name[1:9], combine = F)

p2 <- lapply(p2, function(x){
  x +
    theme(
      aspect.ratio = 1,
      axis.title = element_blank(),
      axis.text = element_blank(),
      axis.ticks = element_blank(),
      axis.line = element_blank(),
      plot.title = element_text(
        size = 8, face = "italic")
    ) +
    NoLegend()
})

p1|wrap_plots(p2, ncol = 3)
```

Here we can see several features associated with high OAR scoring cells (i.e. *PPP1R15A, HSPB1, EIF1, UBC, NMI, DNAJC7, PMAIP1, SAR1A*) as well as one (*B2M*) which is associated with low OAR scoring cells.
