---
title: "Introductory Vignette for OARscRNA"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introductory Vignette for OARscRNA}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = F}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
library(knitr)

hook_output <- knitr::knit_hooks$get("output")
knit_hooks$set(output = function(x, options) {
   lines <- options$output.lines
   if (is.null(lines)) {
     return(hook_output(x, options))  # pass to default hook
   }
   x <- unlist(strsplit(x, "\n"))
   more <- "..."
   if (length(lines)==1) {        # first n lines
     if (length(x) > lines) {
       # truncate the output, but add ....
       x <- c(head(x, lines), more)
     }
   } else {
     x <- c(more, x[lines], more)
   }
   # paste these lines together
   x <- paste(c(x, ""), collapse = "\n")
   hook_output(x, options)
 })

opts_chunk$set(eval=TRUE,
   ## text results
   echo=TRUE,
   results=c('markup', 'asis', 'hold', 'hide')[1],
   ## plots
   fig.path=c('figure', 'figure/minimal-')[1],
   fig.keep=c('high', 'none', 'all', 'first', 'last')[1],
   fig.align=c('center', 'left', 'right', 'default')[1],
   fig.show=c('hold', 'asis', 'animate', 'hide')[1],
   dev=c('pdf', 'png', 'tikz')[2],
   fig.width=5, fig.height=3, #inches
   fig.env=c('figure', 'marginfigure')[1],
   fig.pos=c('', 'h', 't', 'b', 'p', 'H')[1]
   )
```

```{r setup}
library(OAR)
```

# Overview

In this vignette we will walk you through the very basic steps of generating and analyzing OAR scores from your single cell dataset.

There are two possible starting points:

1.  Seurat v5 Object
2.  Gene Expression Matrix

Here we will walk you through steps of the analysis, beginning with a gene expression matrix from plasmacytoid dendritic cell dataset that you can download from our [github repository](https://github.com/davidsanin/OARscRNA/tree/main/vignettes).

## Starting from v5 Seurat Object

### Analysis

We can directly input a Seurat object to examine heterogeneity.

```{r load seurat object}
data_obj <- readRDS(file = "pdcs.rds")
data_obj
```

Once you have the seurat version 5 object loaded, you can run the oar function. This function requires the input data, as well as some information on the number of cores (which is set to 1 by **DEFAULT**).

To determine your computers capacity, you can run `parallelly::availableCores()`. Leave at least 1-2 cores available for other processes running in your computer to prevent crashing. Using multiple cores is recommended as it will go much faster in the identification of the patterns.

```{r run basic oar, output.lines=5}
oar_obj <- oar(data_obj, cores = parallelly::availableCores()-2)
```

We deliberately left most parameters as defaults, however we recommend you explore each to familiarize yourself with their impact on the result. Some critical parameters to consider include:

-   `seurat_v5` confirms the type of data being used. Defaults to `TRUE`.
-   `mismatch` controls whether or not to allow for flexibility in pattern matching. Defaults to `TRUE`.
-   `tolerance` controls whether the mismatch threshold should be determined automatically (between 0.01 and 0.05 hamming distances) or if a user supplied threshold should be used. Defaults to `TRUE`, and can take `TRUE` or a numeric value smaller than 1.
-   `count.filter` controls the minimum percentage of cells expressing any given gene for that gene to be retained in the analysis. Defaults to 1.
-   `blacklisted.genes` allows you to provide a vector of uninformative gene names (i.e. non-coding genes) to exclude from the analysis. Defaults to `NULL`.

The result is a Seurat object with these additional metadata columns:

-   `OARscore` is a metric of heterogeneity in the data, with a higher OAR score highlighting cells with gene expression patterns dissimilar from all others tested.
-   `KW.pvalue` is the p-value generated by Kruskal-Wallis test.
-   `KW.BH.pvalue` is the Benjamini-Hochberg corrected p-value.
-   `pct.missing` is the percent of missing genes in each cell as a fraction of all genes expressed by all cells in the dataset.

Additionally, genes are annotated by which missing data patterns (mdp) they have been included in. This mdp matrix has been added to the assay metadata (found in [oar_obj\@assays](mailto:oar_obj@assays){.email}\$[RNA\@meta.data](mailto:RNA@meta.data){.email}).

### Analysis by Cluster

OAR scores are typically more informative when distinguishing among cells of the same type. When working with a dataset with diverse cell types, it can be helpful to run the function by cluster.

```{r run oar by cluster, output.lines=5, eval = FALSE}
oar_obj_bycluster <- oar_by_cluster(data_obj, cores = parallelly::availableCores()-2)
```

### Visualize Results

Below are a few examples of how you might explore the results of this analysis.

#### OARscore v Percent Missing Scatter Plot

```{r create_scatter_plot, out.width="100%", fig.cap="Scatter Plots"}
library(patchwork)
library(ggplot2)
p1 <- scatter_score_missing(oar_obj, pt.size = 0.5)+theme(legend.position = "none")
p2 <- scatter_score_missing(oar_obj, pt.size = 0.5, group.by = "condition")+theme(legend.position = "none")
p1+p2
```

These plot reveals which clusters have the highest OAR scores, as well as how the score relates to the percent of missing genes. Adjust the `group.by` metric to examine other variables.

#### Missing Data Patterns

In some cases, you might want to explore the patterns that were identified during the analysis to understand what the co-expression gene modules are telling you. First, you can visualize missing data patterns like this:

```{r patterns_found, out.width="50%", fig.cap="Missing Data Pattern Plot"}
p3 <- oar_missing_data_plot(oar_obj)
p3
```

For this dataset, we identified 4 sets of genes: 3 relatively conserved patterns allowing for some mismatch, and a fourth which is a combination of all genes with unique patterns. Each slice of the bar represents a cell, and it is colored based on what percentage of genes from that pattern are expressed in each cell. For example, a pattern of 13 genes is expressed in only about a 10% of cells.

To get a data.frame of the genes within each pattern, use:

```{r get list of genes involved in patterns, eval = FALSE}
mdp <- get_missing_pattern_genes(oar_obj)
```

It is also possible to retrieve missing data patterns when `oar_by_cluster()` is applied. Please see the documentation for `get_missing_pattern_genes()` for more details. Currently, the `oar_missing_data_plot()` is not supported when `oar_by_cluster()` is executed.

#### Score projection with Seurat FeaturePlot

As the scores, p values and sparsity have been added as columns to the `meta.data` of the Seurat object, you can use all the native functions from that package to visualize your results. `Seurat::FeaturePlot()` in particular, is a good starting point to get an overview of how the score is distributed among cells and clusters.

```{r OAR_featureplot, out.width="100%", fig.cap="Feature plot of results"}
library(Seurat)
p4 <- FeaturePlot(
  oar_obj, features = "OARscore", order = T, pt.size = 0.5, 
  min.cutoff = "q40", max.cutoff = "q90")
p4
```

## Starting from a Gene Expression Matrix

The steps to analyse a gene count matrix, with cells as columns and genes as rows, are very similar to what we saw above.

```{r load matrix, output.lines=5}
data <- readRDS(file = "pdcs_matrix.rds")
```

### Analysis

Once you have data loaded, you can run the oar function. In addition to the input data we must specify that we are not starting with a Seurat object by setting `seurat_v5` to `FALSE`.

```{r run OAR on matrix, output.lines=5}
oar_data<- oar(data, seurat_v5 = FALSE, 
               mismatch = TRUE, tolerance = TRUE, 
               cores = parallelly::availableCores()-2)
oar_data
```

This results in a table with the following columns:

-   `OARscore` is a metric of heterogeneity in the data, with a higher OAR score highlighting cells with gene expression patterns dissimilar from all others tested.
-   `KW.pvalue` is the p-value generated by Kruskal-Wallis test.
-   `KW.BH.pvalue` is the Benjamini-Hochberg corrected p-value.
-   `pct.missing` is the percent of missing genes in each cell as a fraction of all genes expressed by all cells in the dataset.

### Visualize Results

We can visualize these OAR scores with a scatter plot against the percent missing genes.

```{r OAR_scatter, out.width="100%", fig.cap="Scatter plot of results"}
p5 <- scatter_score_missing(oar_data, seurat_v5 = F)
p5
```
