---
title: "Step-by-step guide to running OAR"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{detailed_tutorial}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = F}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
library(knitr)

hook_output <- knitr::knit_hooks$get("output")
knit_hooks$set(output = function(x, options) {
   lines <- options$output.lines
   if (is.null(lines)) {
     return(hook_output(x, options))  # pass to default hook
   }
   x <- unlist(strsplit(x, "\n"))
   more <- "..."
   if (length(lines)==1) {        # first n lines
     if (length(x) > lines) {
       # truncate the output, but add ....
       x <- c(head(x, lines), more)
     }
   } else {
     x <- c(more, x[lines], more)
   }
   # paste these lines together
   x <- paste(c(x, ""), collapse = "\n")
   hook_output(x, options)
 })
```

```{r setup}
library(OAR)
```

```{r include=F, echo=F}
#path_to_file_1 <- system.file("extdata", "pdcs.rds", package="OAR")
#path_to_file_2 <- system.file("extdata", "pdcs_matrix.rds", package="OAR")
path_to_file_1 <- "/Users/repository/Git/OARscore/inst/extdata/pdcs.rds"
path_to_file_2 <- "/Users/repository/Git/OARscore/inst/extdata/pdcs_matrix.rds"
```

# Overview

In this vignette we will walk you through the very basic steps of generating and analyzing OARscores from your single cell dataset.

There are two possible starting points: 1. Seurat v5 Object 2. Gene Expression Matrix

Here we will walk you through steps of the analysis, beginning with a gene expression matrix from plasmacytoid dendritic cell dataset that you can download from our [github](https://github.com/davidsanin/OARscRNA/tree/main/inst/extdata).

## Starting from v5 Seurat Object

### Analysis

We can directly input a Seurat object to examine heterogeneity.

```{r load seurat object}
data_obj <- readRDS(file = path_to_file_1)
data_obj
```

Once you have the seurat verwion 5 object loaded, you can run the oar function. This function requires the input data, as well as some information on the number of cores (which is set to 1 by **DEFAULT**).

To determine your computers capacity, you can run parallelly::availableCores(). Leave at least 1-2 cores available for other processes running in your computer to prevent crashing. Using multiple cores is recommened as it will go much faster in the identification of the patterns.

```{r run basic oar, output.lines=5}
oar_obj <- oar(data_obj, cores = 48)
```

We deliberately left most parameters as defaults, however we recommend you explore each to familiarize yourself with their impact on the result. Some critical parameters to consider include:

-   `seurat_v5`: which confirms the type of data being used. Defaults to `TRUE`.
-   `mismatch`: whcih controld whether or not to allow for flexibility in pattern matching. Defaults to `TRUE`.
-   `tolerance`: controls whether the missmatch treshold should be determined automatically (between 0.01 and 0.05 hamming distances) or if a user supplied threshold should be used. Defaults to `TRUE`, and can take `TRUE` or a numeric value smaller than 1.
-   `count.filter`: controls the minimum percentage of cells expressing any given gene for that gene to be retained in the analysis. Defaults to 1.
-   `blacklisted.genes`: Allows you to provide a vector of uninformative gene names (i.e. non-conding genes) to exclude from the analysis. Defaults to `NULL`.

The result is the Seurat object with additional metadata. To the cells metadata (found in [oar_obj\@meta.data](mailto:oar_obj@meta.data){.email}) there will be the following additional columns:

-   `OARscore`: a metric of heterogeneity in the data, with a higher OAR score highlighting cells with gene expression patterns dissimilar from all others tested.
-   `KW.pvalue`: P-value generated by Kruskal-Wallis test.
-   `KW.BH.pvalue`: Benjamini-Hochberg corrected p-value.
-   `pct.missing`: the percent of missing genes in each cell as a fraction of all genes expressed by all cells in the dataset.

Additionally, genes are annotated by which missing data patterns (mdp) they have been included in. This mdp matrix has been added to the assay metadata (found in [oar_obj\@assays](mailto:oar_obj@assays){.email}\$[RNA\@meta.data](mailto:RNA@meta.data){.email}).

### Analysis by Cluster

OARscores are typically more powerful when distinguishing amongst cells as the same type, as heterogeneity in gene patterns are likely from a biological phenomenon other than cell type. When working with a dataset with diverse cell types, it can be helpful to run the function by cluster.

```{r run oar by cluster, output.lines=5}
oar_obj_bycluster <- oar_by_cluster(data_obj, cores = 48)
```

### Visualize Results

There are several way to visualize your resulting seurat object:

#### OARscore v Percent Missing Scatter Plot

```{r create scatter plot, out.width="100%", fig.cap="Scatter Plots"}
library(patchwork)
library(ggplot2)
p1 <- scatter_score_missing(oar_obj)+theme(legend.position = "bottom")
p2 <- scatter_score_missing(oar_obj,group.by = "condition")+theme(legend.position = "bottom")
p1+p2
```

These plot reveals which clusters have the highest OAR scores, as well as how the OARscore relates to the percent of missing genes. Adjust the group.by metric to examine other variables.

#### Graphic of Missing Data Patterns

Visualize patterns found by OAR:

```{r create visual of patterns found, out.width="100%", fig.cap="Missing Data Pattern Plot"}
oar_missing_data_plot(oar_obj)
```

In this case we have a combination of all unique patterns, and 3 patterns found across cells, with the assigned indices of 9102, 6, and 3. Each slice of the bar represents a cell, and it is colored based on what percentage of genes from that pattern are expressed in each cell. For example, in pattern 9102, about half the cells have some of these 13 genes expressed.

To get a list of the highly variable genes withing each pattern, use:

```{r get list of genes involved in patterns"}
mdp <- get_missing_pattern_genes(oar_obj)
```

#### Seurat Feature Plot

To visualize which cells/clusters have the most heterogeneity:

```{r Examine at OAR score in featureplot, fig.cap="Feature plot of results"}
library(Seurat)
FeaturePlot(
  oar_obj, features = "OARscore", order = T, pt.size = 0.5, 
  min.cutoff = "q40", max.cutoff = "q90")
```

## Starting from a Gene Expression Matrix

The steps to analyse a gene count matrix, with cell as columns and genes as rows, are very similar to what we saw above.

```{r load matrix}
data <- readRDS(file = path_to_file_2)
```

### Analysis

Once you have data loaded, you can run the oar function. This function requires the input data, as well as to be informed whether or not it is a v5 Seurat object.

```{r run OAR on matrix, output.lines=5}
oar_data<- oar(data, seurat_v5 = FALSE, 
               mismatch = TRUE, tolerance = TRUE, 
               cores = 48)
oar_data
```

This results in a table with the following columns:

-   `OARscore`: a metric of heterogeneity in the data, with a higher OAR score highlighting cells with gene expression patterns dissimilar from all others tested.
-   `KW.pvalue`: P-value generated by Kruskal-Wallis test.
-   `KW.BH.pvalue`: Benjamini-Hochberg corrected p-value.
-   `pct.missing`: the percent of missing genes in each cell as a fraction of all genes expressed by all cells in the dataset.

### Visualize Results

We can visualize these OARscores via scatter plot against the percent missing genes.

```{r Scatter plot, out.width="50%", fig.cap="Scatter plot of results"}
scatter_score_missing(oar_data, seurat_v5 = F)
```
