<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Step-by-step guide to running OAR • OAR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Step-by-step guide to running OAR">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">OAR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/introductory_vignette.html">Introductory Vignette for OARscRNA</a></li>
    <li><a class="dropdown-item" href="../articles/detailed_tutorial.html">Step-by-step guide to running OAR</a></li>
    <li><a class="dropdown-item" href="../articles/Gene_expression.html">Gene Expression Analysis based on OAR scores</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Sanin-Lab/OARscRNA/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Step-by-step guide to running OAR</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Sanin-Lab/OARscRNA/blob/main/vignettes/detailed_tutorial.Rmd" class="external-link"><code>vignettes/detailed_tutorial.Rmd</code></a></small>
      <div class="d-none name"><code>detailed_tutorial.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Sanin-Lab/OARscRNA" class="external-link">OAR</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>In this tutorial we will walk you through all the steps of generating
and analyzing OAR scores from your single cell dataset. We will begin
with a Seurat object with plasmacytoid dendritic cells that you can
download from our <a href="https://github.com/Sanin-Lab/OARscRNA/tree/main/vignettes" class="external-link">github
repository</a>.</p>
<p>The process starting from a gene expression matrix is very similar
and you should be able to adapt this workflow to that input. Skip a few
lines down to the <strong>data pre-processing step</strong> to run the
tutorial that way.</p>
<p>At the end, we provide an alternative single line <a href="wrapper-functions-and-result-interpretation">wrapper function</a>
that runs the whole process, and adds the results to the
<code>meta.data</code> slot of a Seurat object. If starting from a
dataset with multiple cell types, we recommend running this process
cluster by cluster, and similarly provide a wrapper function that
accomplishes that.</p>
<div class="section level3">
<h3 id="loading-and-pre-processing-your-data">1. Loading and pre-processing your data<a class="anchor" aria-label="anchor" href="#loading-and-pre-processing-your-data"></a>
</h3>
<p>First, we load a Seurat object that we wish examine and store the
cell barcodes so we can later assign them to our output.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sc.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"pdcs.rds"</span><span class="op">)</span></span>
<span><span class="va">cells</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sc.data</span><span class="op">)</span></span>
<span><span class="va">sc.data</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 14362 features across 1243 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (14362 features, 2000 variable features)</span></span>
<span><span class="co">#&gt;  1 layer present: counts</span></span>
<span><span class="co">#&gt;  2 dimensional reductions calculated: pca, umap</span></span></code></pre></div>
<p>Now lets set a count threshold where we only analyze genes expressed
in at least 1% of cells<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;&lt;em&gt;Filtering genes may reduce the time it takes to run
the analysis.&lt;/em&gt;&lt;/p&gt;"><sup>1</sup></a>. We do not recommend using thresholds
higher than 2%, as information will be lost.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tr</span> <span class="op">=</span> <span class="fl">1</span></span></code></pre></div>
<p>Next, we retrieve the un-normalized count data from the Seurat
object, filter low abundance genes and remove genes that are
uninformative<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;&lt;em&gt;Filtering genes may reduce the time it takes to run
the analysis.&lt;/em&gt;&lt;/p&gt;"><sup>2</sup></a>. The output of this function is a list with
the first element being the processed matrix and the second the list of
genes in that matrix after filtering. We will skip the later for this
tutorial. Missing values in the processed matrix of gene counts are
denoted as <code>NA</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">oar_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/oar_preprocess_data.html">oar_preprocess_data</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">sc.data</span>, tr <span class="op">=</span> <span class="va">tr</span>, </span>
<span>  seurat_v5 <span class="op">=</span> <span class="cn">T</span>, blacklisted.genes <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Extracting count tables"</span></span>
<span><span class="va">oar_data</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;      [,1] [,2] [,3] [,4] [,5]</span></span>
<span><span class="co">#&gt; [1,]   NA    1   NA    1   NA</span></span>
<span><span class="co">#&gt; [2,]   NA   NA   NA   NA   NA</span></span>
<span><span class="co">#&gt; [3,]   NA   NA    1   NA   NA</span></span>
<span><span class="co">#&gt; [4,]   NA   NA   NA   NA   NA</span></span>
<span><span class="va">...</span></span></code></pre></div>
<p>For a full description of each parameter consult the documentation.
Briefly:</p>
<ul>
<li>
<code>seurat_v5</code> confirms the type of data being used.
Defaults to <code>TRUE</code>. If <strong>starting from a data
matrix</strong>, set this to <code>FALSE</code> and continue with the
rest of the tutorial.</li>
<li>
<code>tr</code> controls the minimum percentage of cells expressing
any given gene for that gene to be retained in the analysis. Defaults to
0.01.</li>
<li>
<code>blacklisted.genes</code> allows you to provide a vector of
uninformative gene names (i.e. non-coding genes) to exclude from the
analysis. Defaults to <code>NULL</code>.</li>
</ul>
</div>
<div class="section level3">
<h3 id="identifying-missing-data-patterns">2. Identifying Missing Data Patterns<a class="anchor" aria-label="anchor" href="#identifying-missing-data-patterns"></a>
</h3>
<p>Our scoring relies on identifying co-expressed genes patterns (which
we call missing data patterns) which is a computationally intensive
process. To do this, we can identify genes with minimal mismatch between
them, but we need to first enable R to run in parallel. To determine
your computers capacity, you can run
<code><a href="https://parallelly.futureverse.org/reference/availableCores.html" class="external-link">parallelly::availableCores()</a></code>. Leave at least 1-2 cores
available for other processes running in your computer to prevent
crashing. Using multiple cores is recommended as it will go much faster
in the identification of the patterns.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">parallelly</span><span class="fu">::</span><span class="fu"><a href="https://parallelly.futureverse.org/reference/availableCores.html" class="external-link">availableCores</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; system </span></span>
<span><span class="co">#&gt;      2</span></span></code></pre></div>
<p>Now we can select an appropriate number to set up parallel processing
for this R session. To identify the missing data patterns allowing for
mismatch, we calculate the hamming distance between all pairs of gene
expression vectors. Internally, the gene vectors are converted to 0
(missing) and 1 (observed) values. This function needs to have the
number of cores specified explicitly. We set <code>tolerance</code> to
<code>TRUE</code> to search for a tolerance value between 0.01 and 0.05.
This function will search for a value that does not introduce biases in
the pattern matching, which occurs with tolerance values greater than
0.05. This tolerance value is the fraction of mismatch between gene
vectors that we allow to consider genes as part of the same pattern.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mdp_auto_tol</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/oar_missing_data_patterns.html">oar_missing_data_patterns</a></span><span class="op">(</span></span>
<span>  <span class="va">oar_data</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, tolerance <span class="op">=</span> <span class="cn">T</span>, cores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Calculating Hamming distances between gene vectors using specified cores."</span></span>
<span><span class="co">#&gt; [1] "This operation may take several minutes"</span></span>
<span><span class="co">#&gt; [1] "Scanning for optimal tolerance..."</span></span>
<span><span class="co">#&gt; [1] "Tolerance set to 0.02"</span></span>
<span><span class="co">#&gt; [1] "Identified 2 non-unique missing data patterns"</span></span>
<span><span class="va">...</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">mdp_auto_tol</span><span class="op">)</span></span>
<span><span class="co">#&gt; mdp_auto_tol</span></span>
<span><span class="co">#&gt;       Pattern 1       Pattern 2 Unique patterns </span></span>
<span><span class="co">#&gt;               2             108           11596</span></span></code></pre></div>
<p>If we have previously run the analysis or we wish to tune the
tolerance to a value outside the automatic range, we can choose our own
tolerance and pass it to the function, which will be faster, but may not
capture as many patterns. Values greater than 0 and lower than 0.05 are
recommended.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/oar_missing_data_patterns.html">oar_missing_data_patterns</a></span><span class="op">(</span><span class="va">oar_data</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, tolerance <span class="op">=</span> <span class="fl">0.05</span>, cores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p>In this dataset a tolerance of 0.02 produced the best results, with 3
distinct patters found across cells (plus the combination of all genes
with non-unique patterns). We can visualize these results like this:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/oar_missing_data_plot.html">oar_missing_data_plot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">oar_data</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, mdp <span class="op">=</span> <span class="va">mdp_auto_tol</span>, seurat_v5 <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Generating plots..."</span></span>
<span><span class="va">p3</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figurepatterns_found_mdp-1.png" alt="Missing Data Pattern Plot" width="50%"><p class="caption">
Missing Data Pattern Plot
</p>
</div>
<p>Each slice of the bar represents a cell, and it is colored based on
what percentage of genes from that pattern are expressed in each cell.
For example, a pattern of 13 genes is expressed in only about a 10% of
cells.</p>
</div>
<div class="section level3">
<h3 id="calculate-oar-scores">3. Calculate OAR scores<a class="anchor" aria-label="anchor" href="#calculate-oar-scores"></a>
</h3>
<p>Now we are ready to do a cell-wise estimation of the overlap between
the gene expression distributions in each detected pattern. When these
distributions are distinct, we find that the data is not missing at
random (NOAR), with highly significant p values and an OAR score of less
than 2. Alternatively, when these distributions overlap, gene expression
appears more randomly distributed (OAR), and cells have a larger p value
and an OAR score greater than 2. Critically, our test incorporates both
the gene expression values and the missing data patterns.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">output</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/oar_base.html">oar_base</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">oar_data</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, mdp <span class="op">=</span> <span class="va">mdp_auto_tol</span><span class="op">)</span></span>
<span><span class="va">output</span><span class="op">$</span><span class="va">barcodes</span> <span class="op">=</span> <span class="va">cells</span> <span class="co">#add back our barcodes</span></span>
<span><span class="va">output</span></span>
<span><span class="co">#&gt;           OARscore     KW.pvalue KW.BH.pvalue pct.missing             barcodes</span></span>
<span><span class="co">#&gt; 1     0.1649214920  8.887958e-82 1.318345e-81    79.46352 AAACCTGAGAACAATC-1_1</span></span>
<span><span class="co">#&gt; 2    -0.1359034105  1.753283e-84 3.262472e-84    87.92927 AAACCTGAGGTGCTAG-1_1</span></span>
<span><span class="co">#&gt; 3    -1.0976845874  8.528426e-94 1.514405e-92    85.27251 AAACCTGGTAGCAAAT-1_1</span></span>
<span><span class="co">#&gt; 4    -1.3087967701  5.958183e-96 2.244249e-94    78.79720 AAACGGGGTACCCAAT-1_1</span></span>
<span><span class="va">...</span></span></code></pre></div>
<p>The result is a matrix with these columns:</p>
<ul>
<li>
<code>OARscore</code> is a metric of heterogeneity in the data, with
a higher OAR score highlighting cells with gene expression patterns
dissimilar from all others tested.</li>
<li>
<code>KW.pvalue</code> is the p-value generated by Kruskal-Wallis
test.</li>
<li>
<code>KW.BH.pvalue</code> is the Benjamini-Hochberg corrected
p-value.</li>
<li>
<code>pct.missing</code> is the percent of missing genes in each
cell as a fraction of all genes expressed by all cells in the
dataset.</li>
</ul>
<p>We can now proceed to visualize these results to identify which cells
are most heterogeneous.</p>
</div>
<div class="section level3">
<h3 id="visualize-results">4. Visualize results<a class="anchor" aria-label="anchor" href="#visualize-results"></a>
</h3>
<p>A higher OAR score is a way to prioritize cells for further analysis.
Greater scores indicate distinct cell expression signatures, where high
scoring cells have different gene co-expression patterns compared to all
other cells in the test. In this example, we observe that cells with a
high degree of sparsity, also have a high OAR score</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">output</span>,</span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">pct.missing</span>,</span>
<span>      y <span class="op">=</span> <span class="va">OARscore</span>, </span>
<span>      color <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">KW.BH.pvalue</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="st">"% Missing Data"</span>,</span>
<span>    y <span class="op">=</span> <span class="st">"OAR score"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span></span>
<span>    yintercept <span class="op">=</span> <span class="fl">2</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span>, </span>
<span>    color <span class="op">=</span> <span class="st">"black"</span>, linewidth <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_colour_gradientn</a></span><span class="op">(</span></span>
<span>    colours <span class="op">=</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#000004FF"</span>,<span class="st">"#3B0F70FF"</span>,<span class="st">"#8C2981FF"</span>,</span>
<span>                    <span class="st">"#DE4968FF"</span>, <span class="st">"#FE9F6DFF"</span>,<span class="st">"#FCFDBFFF"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    na.value <span class="op">=</span> <span class="st">"transparent"</span>, </span>
<span>    name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="op">-</span><span class="va">Log</span><span class="op">[</span><span class="fl">10</span><span class="op">]</span><span class="op">~</span><span class="va">BH.pval</span><span class="op">)</span>,</span>
<span>    guide <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_colourbar.html" class="external-link">guide_colorbar</a></span><span class="op">(</span></span>
<span>      frame.colour <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>      frame.linewidth <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>      ticks.linewidth <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>      ticks.colour <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>    aspect.ratio <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    panel.grid <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    panel.background <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    axis.line <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_line</a></span><span class="op">(</span>linewidth <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span>,</span>
<span>    legend.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">6</span><span class="op">)</span>, </span>
<span>    legend.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>,</span>
<span>    legend.key.size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">3</span>, <span class="st">"mm"</span><span class="op">)</span>,</span>
<span>    axis.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">6</span><span class="op">)</span>, </span>
<span>    axis.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>,</span>
<span>    plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p1</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figurecreate_scatter_plot-1.png" alt="Scatter Plots" width="100%"><p class="caption">
Scatter Plots
</p>
</div>
<p>You may recreate this plot using
<code><a href="../reference/scatter_score_missing.html">scatter_score_missing()</a></code>.</p>
<p>In order to fully understand the features in the dataset driving the
high OAR score in some cells, it will be necessary to annotate these
with their biological context. This is more easily achieved within a
Seurat object with the additional experimental details.</p>
</div>
<div class="section level3">
<h3 id="wrapper-functions-and-result-interpretation">5. Wrapper functions and result interpretation<a class="anchor" aria-label="anchor" href="#wrapper-functions-and-result-interpretation"></a>
</h3>
<p>We can run this entire process with a single line. The result is the
Seurat object with the <code>OARscore</code>, <code>KW.pvalue</code>,
<code>KW.BH.pvalue</code> and <code>pct.missing</code> values in the
<code>meta.data</code> slot. Additionally, genes are annotated by which
missing data patterns (mdp) they have been included in
<code>sc.data@assays$RNA@meta.data</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sc.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/oar.html">oar</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sc.data</span>, </span>
<span>               seurat_v5 <span class="op">=</span> <span class="cn">T</span>, count.filter <span class="op">=</span> <span class="fl">1</span>,</span>
<span>               blacklisted.genes <span class="op">=</span> <span class="cn">NULL</span>, suffix <span class="op">=</span> <span class="st">""</span>,</span>
<span>               tolerance <span class="op">=</span> <span class="fl">0.02</span>,</span>
<span>               cores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in oar(data = sc.data, seurat_v5 = T, count.filter = 1, blacklisted.genes = NULL, : Running process in fewer than 2 cores will considerably slow down progress</span></span>
<span><span class="co">#&gt; [1] "Extracting data..."</span></span>
<span><span class="co">#&gt; [1] "Extracting count tables"</span></span>
<span><span class="co">#&gt; [1] "Analysis started on:"</span></span>
<span><span class="co">#&gt; [1] "2025-03-24 23:11:42 UTC"</span></span>
<span><span class="co">#&gt; [1] "Identifying missing data patterns..."</span></span>
<span><span class="va">...</span></span></code></pre></div>
<p>We set these parameters based on our earlier results and adjusted to
the data provided:</p>
<ul>
<li>
<code>seurat_v5</code> confirms the type of data being used.
Defaults to <code>TRUE</code>.</li>
<li>
<code>tolerance</code> controls whether the mismatch threshold
should be determined automatically (between 0.01 and 0.05 hamming
distances) or if a user supplied threshold should be used. Defaults to
<code>TRUE</code>, and can take <code>TRUE</code> or a numeric value
smaller than 1. Here we set it to 0.02 based on our results above.</li>
<li>
<code>count.filter</code> controls the minimum percentage of cells
expressing any given gene for that gene to be retained in the analysis.
Defaults to 1.</li>
<li>
<code>blacklisted.genes</code> Allows you to provide a vector of
uninformative gene names (i.e. non-conding genes) to exclude from the
analysis. Defaults to <code>NULL</code>.</li>
<li>
<code>suffix</code> Allows you to provide a character string to
avoid overwriting previous OAR calculations. Defaults to
<code>NULL</code>.</li>
</ul>
<p>You can use a similar approach to analyse the data cluster by cluster
- recommended when you have multiple cell types. In that case, use
<code><a href="../reference/oar_by_cluster.html">oar_by_cluster()</a></code>.</p>
<div class="section level5">
<h5 id="interpreting-high-oar-scores">Interpreting high OAR scores<a class="anchor" aria-label="anchor" href="#interpreting-high-oar-scores"></a>
</h5>
<p>To explore which cells/clusters have the most heterogeneity, we can
project the scores to our UMAP using <code><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">Seurat::FeaturePlot()</a></code>.
Notice that cells with a high OAR score in this dataset cluster
together. In fact, in this case, high OAR scoring cells represent the
most activated cells in the experiment, as can be seen when we examine
<em>IFNA1</em>, which is expressed in activated plasmacytoid dendritic
cells.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: SeuratObject</span></span>
<span><span class="co">#&gt; Loading required package: sp</span></span>
<span><span class="co">#&gt; 'SeuratObject' was built under R 4.4.0 but the current version is</span></span>
<span><span class="co">#&gt; 4.4.3; it is recomended that you reinstall 'SeuratObject' as the ABI</span></span>
<span><span class="co">#&gt; for R may have changed</span></span>
<span><span class="co">#&gt; 'SeuratObject' was built with package 'Matrix' 1.7.1 but the current</span></span>
<span><span class="co">#&gt; version is 1.7.2; it is recomended that you reinstall 'SeuratObject' as</span></span>
<span><span class="co">#&gt; the ABI for 'Matrix' may have changed</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'SeuratObject'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, t</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span></span>
<span>  <span class="va">sc.data</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"OARscore"</span>,<span class="st">"IFNA1"</span><span class="op">)</span>, order <span class="op">=</span> <span class="cn">T</span>, pt.size <span class="op">=</span> <span class="fl">0.5</span>, </span>
<span>  min.cutoff <span class="op">=</span> <span class="st">"q40"</span>, max.cutoff <span class="op">=</span> <span class="st">"q90"</span>, slot <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span></span>
<span><span class="va">p1</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figurefeatureplot_score-1.png" alt="Feature plot of OAR score" width="100%"><p class="caption">
Feature plot of OAR score
</p>
</div>
<p>Thus, in this dataset we are able to identify highly activated cells
with a cluster agnostic approach. We provide additional tutorials to
identify genes associated with a high OAR score (see:
<code><a href="../articles/Gene_expression.html">vignette("Gene_expression")</a></code>). We have also seen datasets
where high scores are associated with biological variation that may be
removed from the analysis to glean additional insights into the
populations of cells studied. These can be achieved by passing
<code>OARscore</code> to <code>vars.to.regress</code> in
<code><a href="https://satijalab.org/seurat/reference/SCTransform.html" class="external-link">Seurat::SCTransform()</a></code>.</p>
</div>
</div>
</div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by David Sanin, Ruizhe Chen, Hannah Moore, Dmitrijs Lvovs, Paul Gueguen.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
