<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Gene Expression Analysis based on OAR scores • OAR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Gene Expression Analysis based on OAR scores">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">OAR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/introductory_vignette.html">Introductory Vignette for OARscRNA</a></li>
    <li><a class="dropdown-item" href="../articles/detailed_tutorial.html">Step-by-step guide to running OAR</a></li>
    <li><a class="dropdown-item" href="../articles/OAR_Factor.html">Running OAR in a dataset split by Factors</a></li>
    <li><a class="dropdown-item" href="../articles/Gene_expression.html">Gene Expression Analysis based on OAR scores</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Sanin-Lab/OARscRNA/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Gene Expression Analysis based on OAR scores</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Sanin-Lab/OARscRNA/blob/main/vignettes/Gene_expression.Rmd" class="external-link"><code>vignettes/Gene_expression.Rmd</code></a></small>
      <div class="d-none name"><code>Gene_expression.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Sanin-Lab/OARscRNA" class="external-link">OAR</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>In this vignette we will explore what genes are associated with
increased OAR scores. We will use our toy plasmacytoid dendritic cell
dataset that you can download from our <a href="https://github.com/Sanin-Lab/OARscRNA/tree/main/vignettes" class="external-link">github
repository</a>.</p>
<div class="section level3">
<h3 id="calculating-oar-scores">1. Calculating OAR scores<a class="anchor" aria-label="anchor" href="#calculating-oar-scores"></a>
</h3>
<p>We can calculate OAR scores with a single line of code. We recommend
you follow the introductory tutorials for a full description of this
function. The result is a Seurat object with <code>OARscore</code>,
<code>KW.pvalue</code>, <code>KW.BH.pvalue</code> and
<code>pct.missing</code> values in the <code>meta.data</code> slot.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: SeuratObject</span></span>
<span><span class="co">#&gt; Loading required package: sp</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'SeuratObject'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, t</span></span>
<span><span class="va">sc.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/oar.html">oar</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"pdcs.rds"</span><span class="op">)</span>, </span>
<span>               seurat_v5 <span class="op">=</span> <span class="cn">T</span>, count.filter <span class="op">=</span> <span class="fl">1</span>,</span>
<span>               blacklisted.genes <span class="op">=</span> <span class="cn">NULL</span>, suffix <span class="op">=</span> <span class="st">""</span>,</span>
<span>               cores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in oar(data = readRDS(file = "pdcs.rds"), seurat_v5 = T, count.filter = 1, : Running process in fewer than 2 cores will considerably slow down progress</span></span>
<span><span class="co">#&gt; [1] "Extracting data..."</span></span>
<span><span class="co">#&gt; [1] "Extracting count tables"</span></span>
<span><span class="co">#&gt; [1] "Analysis started on:"</span></span>
<span><span class="co">#&gt; [1] "2025-06-02 19:41:38 UTC"</span></span>
<span><span class="co">#&gt; [1] "Identifying missing data patterns..."</span></span>
<span><span class="va">...</span></span></code></pre></div>
<p>To explore which cells/clusters have the most heterogeneity, we can
visualize the results in a UMAP projection.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span></span>
<span>  <span class="va">sc.data</span>, features <span class="op">=</span> <span class="st">"OARscore"</span>, order <span class="op">=</span> <span class="cn">T</span>, pt.size <span class="op">=</span> <span class="fl">0.5</span>, </span>
<span>  min.cutoff <span class="op">=</span> <span class="st">"q40"</span>, max.cutoff <span class="op">=</span> <span class="st">"q90"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: The `slot` argument of `FetchData()` is deprecated as of SeuratObject 5.0.0.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Please use the `layer` argument instead.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> The deprecated feature was likely used in the <span style="color: #0000BB;">Seurat</span> package.</span></span>
<span><span class="co">#&gt;   Please report the issue at <span style="color: #0000BB; font-style: italic;">&lt;https://github.com/satijalab/seurat/issues&gt;</span>.</span></span>
<span><span class="co">#&gt; <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">generated.</span></span></span>
<span><span class="va">p1</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figurefeatureplot_score_deg-1.png" alt="Feature plot of OAR score" width="100%"><p class="caption">
Feature plot of OAR score
</p>
</div>
<p>Notice that cells with a High OAR score in this dataset cluster
together.</p>
</div>
<div class="section level3">
<h3 id="identify-genes-associated-with-oar-scores">2. Identify genes associated with OAR scores<a class="anchor" aria-label="anchor" href="#identify-genes-associated-with-oar-scores"></a>
</h3>
<p>In a typical scRNAseq analysis workflow, differential gene expression
is calculated based on averaged counts across identified clusters.
Although highly informative, this approach fundamentally looses the
single cell resolution achieved through this technology. As an
alternative to using cluster labels to model gene expression, which
invites biases associated with identifying clusters, we propose to model
gene expression based instead on the value of the OAR score. We
implement this approach using glmGamPoi<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;glmGapPoi: Fit Gamma-Poisson Generalized Linear Models
Reliably. &lt;a href="https://github.com/const-ae/glmGamPoi" class="external-link"&gt;git&lt;/a&gt;&lt;/p&gt;'><sup>1</sup></a>, which has superior
performance to other differential gene expression methods in terms of
speed and accuracy. Scaling factors are calculated using
<code><a href="https://rdrr.io/pkg/scran/man/computeSumFactors.html" class="external-link">scran::calculateSumFactors()</a></code>, which benefits from parallel
computing.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">oar_deg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/oar_deg.html">oar_deg</a></span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="va">sc.data</span>, seurat_v5 <span class="op">=</span> <span class="cn">T</span>, score.name <span class="op">=</span> <span class="st">"OARscore"</span>, </span>
<span>    score <span class="op">=</span> <span class="cn">NULL</span>, count.filter <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    blacklisted.genes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>    splines <span class="op">=</span> <span class="cn">TRUE</span>, degrees.freedom <span class="op">=</span> <span class="fl">5</span>,</span>
<span>    auto.threshold <span class="op">=</span> <span class="cn">TRUE</span>, custom.tr <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Analysis started on:"</span></span>
<span><span class="co">#&gt; [1] "2025-06-02 19:41:46 UTC"</span></span>
<span><span class="co">#&gt; [1] "Extracting count tables"</span></span>
<span><span class="co">#&gt; [1] "FDR threshold set to:"</span></span>
<span><span class="co">#&gt; [1] 1e-20</span></span>
<span><span class="co">#&gt; Warning in oar_deg(data = sc.data, seurat_v5 = T, score.name = "OARscore", : Using splines increases calculation time by 3-5x</span></span>
<span><span class="co">#&gt; [1] "Analysis completed at:"</span></span>
<span><span class="co">#&gt; [1] "2025-06-02 19:42:36 UTC"</span></span></code></pre></div>
<p>A brief explanation of these parameters is presented here:</p>
<ul>
<li>
<code>data</code> may be an raw matrix of counts or a fully
processed Seurat object with an OAR score calculated, as indicated by
the <code>seurat_v5</code> parameter. The name of the score should be
indicated in the <code>score.name</code> parameter.</li>
<li>
<code>score</code> only needs to be provided if starting from a
count matrix and should be a vector of OAR scores in the same order as
the columns of the count matrix.</li>
<li>
<code>count.filter</code> controls the minimum percentage of cells
expressing any given gene for that gene to be retained in the analysis.
Defaults to 1.</li>
<li>
<code>blacklisted.genes</code>, Allows you to provide a vector of
uninformative gene names (i.e. non-coding genes) to exclude from the
analysis. <strong>This analysis automatically excludes all ribosomal
genes</strong>. Defaults to <code>NULL</code>.</li>
<li>
<code>splines</code> and <code>degrees.freedom</code>, control
whether splines should be fit based on the OAR score (and how many, as
set by <code>degrees.freedom</code>) to capture non-linear associations
between scores and gene expression. We recommend setting
<code>splines</code> to <code>TRUE</code> and
<code>degrees.freedom</code> between 3-5. If <code>splines</code> is set
to <code>FALSE</code>, only monotonic/linear associations will be
retrieved.</li>
<li>
<code>auto.threshold</code> and <code>custom.tr</code> control the
false discovery rate (<strong>FDR</strong>) threshold to consider a gene
significantly associated with the OAR score. <code>auto.threshold</code>
set to <code>TRUE</code> (<strong>recommended</strong>) estimates this
threshold based on the number of cells tested. Typically, given the high
number of observations, p values calculated based on individual cells
tend to be highly significant, even where little association is observed
between gene expression and the predictive variable. If
<code>auto.threshold</code> set to <code>FALSE</code>, then a custom p
value threshold should be provided with <code>custom.tr</code>. Setting
this value to typical ranges (i.e. 0.05-0.01) will likely result in
spurious associations.</li>
</ul>
<p>The results of the analysis is a matrix with the regulated and
filtered genes:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">oar_deg</span></span>
<span><span class="co">#&gt;           name         pval     adj_pval f_statistic df1      df2 lfc</span></span>
<span><span class="co">#&gt; 1        HSPB1 1.420100e-87 1.655695e-83    98.24009   5 1262.951  NA</span></span>
<span><span class="co">#&gt; 2     PPP1R15A 1.884017e-67 1.098288e-63    73.25076   5 1262.951  NA</span></span>
<span><span class="co">#&gt; 3      HLA-DRA 1.599269e-61 6.215294e-58    66.22267   5 1262.951  NA</span></span>
<span><span class="co">#&gt; 4          UBC 1.854397e-60 5.405103e-57    64.97655   5 1262.951  NA</span></span>
<span><span class="co">#&gt; 5     HLA-DRB1 5.356963e-60 1.249137e-56    64.43856   5 1262.951  NA</span></span>
<span><span class="co">#&gt; 6     HSP90AB1 3.049325e-56 5.925347e-53    60.08569   5 1262.951  NA</span></span>
<span><span class="co">#&gt; 7     HSP90AA1 1.649319e-50 2.747058e-47    53.54926   5 1262.951  NA</span></span>
<span><span class="co">#&gt; 8     HLA-DPA1 5.614897e-50 8.183010e-47    52.94923   5 1262.951  NA</span></span>
<span><span class="co">#&gt; 9          UBB 2.792879e-49 3.618020e-46    52.16515   5 1262.951  NA</span></span>
<span><span class="va">...</span></span></code></pre></div>
<p>For a full description of each column, visit the
<code>glmGamPoi</code> documentation. For our purposes, the
<code>name</code> and <code>adj_pval</code> columns (corresponding to
genes and FDR, respectively) are the most important results.</p>
</div>
<div class="section level3">
<h3 id="visualize-results">3. Visualize results<a class="anchor" aria-label="anchor" href="#visualize-results"></a>
</h3>
<p>Overall, this analysis identifies 135 genes significantly associated
with OAR scores. If we plot the top 9 most significant genes alongside
our previously calculated OAR scores, we can observe that we identified
several candidates for further analysis:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span></span>
<span>  <span class="va">sc.data</span>, features <span class="op">=</span> <span class="st">"OARscore"</span>, order <span class="op">=</span> <span class="cn">T</span>, pt.size <span class="op">=</span> <span class="fl">0.5</span>, </span>
<span>  min.cutoff <span class="op">=</span> <span class="st">"q40"</span>, max.cutoff <span class="op">=</span> <span class="st">"q90"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"OAR score"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>      aspect.ratio <span class="op">=</span> <span class="fl">1</span>,</span>
<span>      legend.position <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>      axis.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>      axis.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>      axis.ticks <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>      axis.line <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>      plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">8</span>, face <span class="op">=</span> <span class="st">"plain"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">sc.data</span>, pt.size <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  slot <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>  min.cutoff <span class="op">=</span> <span class="st">"q40"</span>, max.cutoff <span class="op">=</span> <span class="st">"q90"</span>,</span>
<span>  features <span class="op">=</span> <span class="va">oar_deg</span><span class="op">$</span><span class="va">name</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">9</span><span class="op">]</span>, combine <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">p2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">x</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>      aspect.ratio <span class="op">=</span> <span class="fl">1</span>,</span>
<span>      axis.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>      axis.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>      axis.ticks <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>      axis.line <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>      plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span></span>
<span>        size <span class="op">=</span> <span class="fl">8</span>, face <span class="op">=</span> <span class="st">"italic"</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span><span class="op">|</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">p2</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figurefeatureplot_comparison-1.png" alt="Genes associated with OAR score" width="100%"><p class="caption">
Genes associated with OAR score
</p>
</div>
<p>Here we can see several features associated with high OAR scoring
cells (i.e. <em>PPP1R15A, HSPB1, EIF1, UBC, NMI, DNAJC7, PMAIP1,
SAR1A</em>) as well as one (<em>B2M</em>) which is associated with low
OAR scoring cells.</p>
</div>
</div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by David Sanin, Ruizhe Chen, Hannah Moore, Dmitrijs Lvovs, Paul Gueguen.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
