<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Introductory Vignette for OARscRNA â€¢ OAR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Introductory Vignette for OARscRNA">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">OAR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/introductory_vignette.html">Introductory Vignette for OARscRNA</a></li>
    <li><a class="dropdown-item" href="../articles/detailed_tutorial.html">Step-by-step guide to running OAR</a></li>
    <li><a class="dropdown-item" href="../articles/Gene_expression.html">Gene Expression Analysis based on OAR scores</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Sanin-Lab/OARscRNA/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Introductory Vignette for OARscRNA</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Sanin-Lab/OARscRNA/blob/main/vignettes/introductory_vignette.Rmd" class="external-link"><code>vignettes/introductory_vignette.Rmd</code></a></small>
      <div class="d-none name"><code>introductory_vignette.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Sanin-Lab/OARscRNA" class="external-link">OAR</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Error in get(paste0(generic, ".", class), envir = get_method_env()) : </span></span>
<span><span class="co">#&gt;   object 'type_sum.accel' not found</span></span></code></pre></div>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>In this vignette we will walk you through the very basic steps of
generating and analyzing OAR scores from your single cell dataset.</p>
<p>There are two possible starting points:</p>
<ol style="list-style-type: decimal">
<li>Seurat v5 Object</li>
<li>Gene Expression Matrix</li>
</ol>
<p>Here we will walk you through steps of the analysis, beginning with a
gene expression matrix from plasmacytoid dendritic cell dataset that you
can download from our <a href="https://github.com/davidsanin/OARscRNA/tree/main/vignettes" class="external-link">github
repository</a>.</p>
<div class="section level3">
<h3 id="starting-from-v5-seurat-object">Starting from v5 Seurat Object<a class="anchor" aria-label="anchor" href="#starting-from-v5-seurat-object"></a>
</h3>
<div class="section level4">
<h4 id="analysis">Analysis<a class="anchor" aria-label="anchor" href="#analysis"></a>
</h4>
<p>We can directly input a Seurat object to examine heterogeneity.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"pdcs.rds"</span><span class="op">)</span></span>
<span><span class="va">data_obj</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 14362 features across 1243 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (14362 features, 2000 variable features)</span></span>
<span><span class="co">#&gt;  1 layer present: counts</span></span>
<span><span class="co">#&gt;  2 dimensional reductions calculated: pca, umap</span></span></code></pre></div>
<p>Once you have the seurat version 5 object loaded, you can run the oar
function. This function requires the input data, as well as some
information on the number of cores (which is set to 1 by
<strong>DEFAULT</strong>).</p>
<p>To determine your computers capacity, you can run
<code><a href="https://parallelly.futureverse.org/reference/availableCores.html" class="external-link">parallelly::availableCores()</a></code>. Leave at least 1-2 cores
available for other processes running in your computer to prevent
crashing. Using multiple cores is recommended as it will go much faster
in the identification of the patterns.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">oar_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/oar.html">oar</a></span><span class="op">(</span><span class="va">data_obj</span>, cores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in oar(data_obj, cores = 1): Running process in fewer than 2 cores will considerably slow down progress</span></span>
<span><span class="co">#&gt; [1] "Extracting data..."</span></span>
<span><span class="co">#&gt; [1] "Extracting count tables"</span></span>
<span><span class="co">#&gt; [1] "Analysis started on:"</span></span>
<span><span class="co">#&gt; [1] "2024-12-18 18:26:20 UTC"</span></span>
<span><span class="co">#&gt; [1] "Identifying missing data patterns..."</span></span>
<span><span class="va">...</span></span></code></pre></div>
<p>We deliberately left most parameters as defaults, however we
recommend you explore each to familiarize yourself with their impact on
the result. Some critical parameters to consider include:</p>
<ul>
<li>
<code>seurat_v5</code> confirms the type of data being used.
Defaults to <code>TRUE</code>.</li>
<li>
<code>mismatch</code> controls whether or not to allow for
flexibility in pattern matching. Defaults to <code>TRUE</code>.</li>
<li>
<code>tolerance</code> controls whether the mismatch threshold
should be determined automatically (between 0.01 and 0.05 hamming
distances) or if a user supplied threshold should be used. Defaults to
<code>TRUE</code>, and can take <code>TRUE</code> or a numeric value
smaller than 1.</li>
<li>
<code>count.filter</code> controls the minimum percentage of cells
expressing any given gene for that gene to be retained in the analysis.
Defaults to 1.</li>
<li>
<code>blacklisted.genes</code> allows you to provide a vector of
uninformative gene names (i.e.Â non-coding genes) to exclude from the
analysis. Defaults to <code>NULL</code>.</li>
</ul>
<p>The result is a Seurat object with these additional metadata
columns:</p>
<ul>
<li>
<code>OARscore</code> is a metric of heterogeneity in the data, with
a higher OAR score highlighting cells with gene expression patterns
dissimilar from all others tested.</li>
<li>
<code>KW.pvalue</code> is the p-value generated by Kruskal-Wallis
test.</li>
<li>
<code>KW.BH.pvalue</code> is the Benjamini-Hochberg corrected
p-value.</li>
<li>
<code>pct.missing</code> is the percent of missing genes in each
cell as a fraction of all genes expressed by all cells in the
dataset.</li>
</ul>
<p>Additionally, genes are annotated by which missing data patterns
(mdp) they have been included in. This mdp matrix has been added to the
assay metadata (found in <a href="mailto:oar_obj@assays" class="email">oar_obj@assays</a>$<a href="mailto:RNA@meta.data" class="email">RNA@meta.data</a>).</p>
</div>
<div class="section level4">
<h4 id="analysis-by-cluster">Analysis by Cluster<a class="anchor" aria-label="anchor" href="#analysis-by-cluster"></a>
</h4>
<p>OAR scores are typically more informative when distinguishing among
cells of the same type. When working with a dataset with diverse cell
types, it can be helpful to run the function by cluster.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">oar_obj_bycluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/oar_by_cluster.html">oar_by_cluster</a></span><span class="op">(</span><span class="va">data_obj</span>, cores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="visualize-results">Visualize Results<a class="anchor" aria-label="anchor" href="#visualize-results"></a>
</h4>
<p>Below are a few examples of how you might explore the results of this
analysis.</p>
<div class="section level5">
<h5 id="oarscore-v-percent-missing-scatter-plot">OARscore v Percent Missing Scatter Plot<a class="anchor" aria-label="anchor" href="#oarscore-v-percent-missing-scatter-plot"></a>
</h5>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scatter_score_missing.html">scatter_score_missing</a></span><span class="op">(</span><span class="va">oar_obj</span>, pt.size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scatter_score_missing.html">scatter_score_missing</a></span><span class="op">(</span><span class="va">oar_obj</span>, pt.size <span class="op">=</span> <span class="fl">0.5</span>, group.by <span class="op">=</span> <span class="st">"condition"</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span><span class="va">p1</span><span class="op">+</span><span class="va">p2</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figurecreate_scatter_plot-1.png" alt="Scatter Plots" width="100%"><p class="caption">
Scatter Plots
</p>
</div>
<p>These plot reveals which clusters have the highest OAR scores, as
well as how the score relates to the percent of missing genes. Adjust
the <code>group.by</code> metric to examine other variables.</p>
</div>
<div class="section level5">
<h5 id="missing-data-patterns">Missing Data Patterns<a class="anchor" aria-label="anchor" href="#missing-data-patterns"></a>
</h5>
<p>In some cases, you might want to explore the patterns that were
identified during the analysis to understand what the co-expression gene
modules are telling you. First, you can visualize missing data patterns
like this:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/oar_missing_data_plot.html">oar_missing_data_plot</a></span><span class="op">(</span><span class="va">oar_obj</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Extracting count tables"</span></span>
<span><span class="co">#&gt; [1] "Generating plots..."</span></span>
<span><span class="va">p3</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figurepatterns_found-1.png" alt="Missing Data Pattern Plot" width="50%"><p class="caption">
Missing Data Pattern Plot
</p>
</div>
<p>For this dataset, we identified 4 sets of genes: 3 relatively
conserved patterns allowing for some mismatch, and a fourth which is a
combination of all genes with unique patterns. Each slice of the bar
represents a cell, and it is colored based on what percentage of genes
from that pattern are expressed in each cell. For example, a pattern of
13 genes is expressed in only about a 10% of cells.</p>
<p>To get a data.frame of the genes within each pattern, use:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mdp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_missing_pattern_genes.html">get_missing_pattern_genes</a></span><span class="op">(</span><span class="va">oar_obj</span><span class="op">)</span></span></code></pre></div>
<p>It is also possible to retrieve missing data patterns when
<code><a href="../reference/oar_by_cluster.html">oar_by_cluster()</a></code> is applied. Please see the documentation
for <code><a href="../reference/get_missing_pattern_genes.html">get_missing_pattern_genes()</a></code> for more details.
Currently, the <code><a href="../reference/oar_missing_data_plot.html">oar_missing_data_plot()</a></code> is not supported
when <code><a href="../reference/oar_by_cluster.html">oar_by_cluster()</a></code> is executed.</p>
</div>
<div class="section level5">
<h5 id="score-projection-with-seurat-featureplot">Score projection with Seurat FeaturePlot<a class="anchor" aria-label="anchor" href="#score-projection-with-seurat-featureplot"></a>
</h5>
<p>As the scores, p values and sparsity have been added as columns to
the <code>meta.data</code> of the Seurat object, you can use all the
native functions from that package to visualize your results.
<code><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">Seurat::FeaturePlot()</a></code> in particular, is a good starting
point to get an overview of how the score is distributed among cells and
clusters.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: SeuratObject</span></span>
<span><span class="co">#&gt; Loading required package: sp</span></span>
<span><span class="co">#&gt; 'SeuratObject' was built under R 4.4.0 but the current version is</span></span>
<span><span class="co">#&gt; 4.4.2; it is recomended that you reinstall 'SeuratObject' as the ABI</span></span>
<span><span class="co">#&gt; for R may have changed</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'SeuratObject'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, t</span></span>
<span><span class="va">p4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span></span>
<span>  <span class="va">oar_obj</span>, features <span class="op">=</span> <span class="st">"OARscore"</span>, order <span class="op">=</span> <span class="cn">T</span>, pt.size <span class="op">=</span> <span class="fl">0.5</span>, </span>
<span>  min.cutoff <span class="op">=</span> <span class="st">"q40"</span>, max.cutoff <span class="op">=</span> <span class="st">"q90"</span><span class="op">)</span></span>
<span><span class="va">p4</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figureOAR_featureplot-1.png" alt="Feature plot of results" width="100%"><p class="caption">
Feature plot of results
</p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="starting-from-a-gene-expression-matrix">Starting from a Gene Expression Matrix<a class="anchor" aria-label="anchor" href="#starting-from-a-gene-expression-matrix"></a>
</h3>
<p>The steps to analyse a gene count matrix, with cells as columns and
genes as rows, are very similar to what we saw above.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"pdcs_matrix.rds"</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="analysis-1">Analysis<a class="anchor" aria-label="anchor" href="#analysis-1"></a>
</h4>
<p>Once you have data loaded, you can run the oar function. In addition
to the input data we must specify that we are not starting with a Seurat
object by setting <code>seurat_v5</code> to <code>FALSE</code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">oar_data</span><span class="op">&lt;-</span> <span class="fu"><a href="../reference/oar.html">oar</a></span><span class="op">(</span><span class="va">data</span>, seurat_v5 <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>               mismatch <span class="op">=</span> <span class="cn">TRUE</span>, tolerance <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>               cores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in oar(data, seurat_v5 = FALSE, mismatch = TRUE, tolerance = TRUE, : Running process in fewer than 2 cores will considerably slow down progress</span></span>
<span><span class="co">#&gt; [1] "Extracting data..."</span></span>
<span><span class="co">#&gt; [1] "Analysis started on:"</span></span>
<span><span class="co">#&gt; [1] "2024-12-18 18:31:28 UTC"</span></span>
<span><span class="co">#&gt; [1] "Identifying missing data patterns..."</span></span>
<span><span class="co">#&gt; [1] "Calculating Hamming distances between gene vectors using specified cores."</span></span>
<span><span class="va">...</span></span>
<span><span class="va">oar_data</span></span>
<span><span class="co">#&gt;          OARscore     KW.pvalue  KW.BH.pvalue pct.missing</span></span>
<span><span class="co">#&gt; 1     0.282688134 8.168498e-107 1.091768e-106    79.46352</span></span>
<span><span class="co">#&gt; 2     0.914667119  2.961274e-97  3.405055e-97    87.92927</span></span>
<span><span class="co">#&gt; 3    -0.301502371 8.795812e-116 1.828293e-115    85.27251</span></span>
<span><span class="co">#&gt; 4    -1.014834292 1.215944e-127 3.514926e-126    78.79720</span></span>
<span><span class="va">...</span></span></code></pre></div>
<p>This results in a table with the following columns:</p>
<ul>
<li>
<code>OARscore</code> is a metric of heterogeneity in the data, with
a higher OAR score highlighting cells with gene expression patterns
dissimilar from all others tested.</li>
<li>
<code>KW.pvalue</code> is the p-value generated by Kruskal-Wallis
test.</li>
<li>
<code>KW.BH.pvalue</code> is the Benjamini-Hochberg corrected
p-value.</li>
<li>
<code>pct.missing</code> is the percent of missing genes in each
cell as a fraction of all genes expressed by all cells in the
dataset.</li>
</ul>
</div>
<div class="section level4">
<h4 id="visualize-results-1">Visualize Results<a class="anchor" aria-label="anchor" href="#visualize-results-1"></a>
</h4>
<p>We can visualize these OAR scores with a scatter plot against the
percent missing genes.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p5</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scatter_score_missing.html">scatter_score_missing</a></span><span class="op">(</span><span class="va">oar_data</span>, seurat_v5 <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="va">p5</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figureOAR_scatter-1.png" alt="Scatter plot of results" width="100%"><p class="caption">
Scatter plot of results
</p>
</div>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by David Sanin, Ruizhe Chen, Hannah Moore, Dmitrijs Lvovs, Paul Gueguen.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
