<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Running OAR in a dataset split by Factors • OAR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Running OAR in a dataset split by Factors">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">OAR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/introductory_vignette.html">Introductory Vignette for OARscRNA</a></li>
    <li><a class="dropdown-item" href="../articles/detailed_tutorial.html">Step-by-step guide to running OAR</a></li>
    <li><a class="dropdown-item" href="../articles/OAR_Factor.html">Running OAR in a dataset split by Factors</a></li>
    <li><a class="dropdown-item" href="../articles/Gene_expression.html">Gene Expression Analysis based on OAR scores</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Sanin-Lab/OARscRNA/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Running OAR in a dataset split by Factors</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Sanin-Lab/OARscRNA/blob/main/vignettes/OAR_Factor.Rmd" class="external-link"><code>vignettes/OAR_Factor.Rmd</code></a></small>
      <div class="d-none name"><code>OAR_Factor.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Sanin-Lab/OARscRNA" class="external-link">OAR</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>In this tutorial we will walk you through generating OAR scores from
your single cell dataset separated by a factor in your data. We will
begin with a Seurat object with <em>Dictyostelium discoideum</em> cells
(dicty) that you can download from our <a href="#id_0">github
repository</a>.</p>
<p>This approach is usually recommended when you have multiple cell
types in your dataset or your experimental conditions introduce dramatic
shifts in gene expression patterns. Our approach works best to detect
heterogeneity within a set of cells that you expect to be somewhat
similar. Consequently, in situations like the ones outlined above, it
makes more sense to split the data by a factor before running the
analysis.</p>
<p>First, we will take a look at how the test performs when the data is
not split. Later we will run the analysis split by factors. As always,
we begin by loading the seurat object with the dataset we are interested
in.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: SeuratObject</span></span>
<span><span class="co">#&gt; Loading required package: sp</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'SeuratObject'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, t</span></span>
<span><span class="va">sc.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"dicty.rds"</span><span class="op">)</span></span>
<span><span class="va">sc.data</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 13206 features across 5000 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (13206 features, 2000 variable features)</span></span>
<span><span class="co">#&gt;  1 layer present: counts</span></span>
<span><span class="co">#&gt;  1 dimensional reduction calculated: umap</span></span></code></pre></div>
<div class="section level3">
<h3 id="running-the-full-test">1. Running the full test<a class="anchor" aria-label="anchor" href="#running-the-full-test"></a>
</h3>
<p>We can run the full test entire process with a single line. For a
description of all these parameters see our other vignettes. The result
is the Seurat object with the <code>OARscore</code>,
<code>KW.pvalue</code>, <code>KW.BH.pvalue</code> and
<code>pct.missing</code> values in the <code>meta.data</code> slot.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sc.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/oar.html">oar</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sc.data</span>, </span>
<span>               seurat_v5 <span class="op">=</span> <span class="cn">T</span>, count.filter <span class="op">=</span> <span class="fl">1</span>,</span>
<span>               blacklisted.genes <span class="op">=</span> <span class="cn">NULL</span>, suffix <span class="op">=</span> <span class="st">""</span>,</span>
<span>               store.hamming <span class="op">=</span> <span class="cn">F</span>,</span>
<span>               cores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in oar(data = sc.data, seurat_v5 = T, count.filter = 1, blacklisted.genes = NULL, : Running process in fewer than 2 cores will considerably slow down progress</span></span>
<span><span class="co">#&gt; [1] "Extracting data..."</span></span>
<span><span class="co">#&gt; [1] "Extracting count tables"</span></span>
<span><span class="co">#&gt; [1] "Analysis started on:"</span></span>
<span><span class="co">#&gt; [1] "2025-05-23 15:41:15 UTC"</span></span>
<span><span class="co">#&gt; [1] "Identifying missing data patterns..."</span></span>
<span><span class="va">...</span></span></code></pre></div>
<p>To explore the results we can visualize <code>OARscore</code> vs
<code>pct.missing</code> values, coloring the cells based on some of the
attributes in our Seurat object meta.data.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scatter_score_missing.html">scatter_score_missing</a></span><span class="op">(</span><span class="va">sc.data</span>, pt.size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scatter_score_missing.html">scatter_score_missing</a></span><span class="op">(</span><span class="va">sc.data</span>, pt.size <span class="op">=</span> <span class="fl">0.5</span>, group.by <span class="op">=</span> <span class="st">"group"</span><span class="op">)</span></span>
<span><span class="va">p1</span><span class="op">+</span><span class="va">p2</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figurecreate_scatter_plot_condition-1.png" alt="Scatter Plots by Factors" width="100%"><p class="caption">
Scatter Plots by Factors
</p>
</div>
<p>Here we can see that the underlying biological conditions are
resulting in very different sets of OARscores. In particular, 10h of
starvation, induces a completely separate profile of values. For that
reason, running the test separating the data by this biological variable
will allow us to prioritize cells within each condition for further
analysis.</p>
</div>
<div class="section level3">
<h3 id="analysis-by-factor">2. Analysis by Factor<a class="anchor" aria-label="anchor" href="#analysis-by-factor"></a>
</h3>
<p>OAR scores are typically more informative when distinguishing among
cells of the same type. When working with a dataset with diverse cell
types or dramatically affected by a biological variable, it can be
helpful to split the data by that factor and run the test independently.
This can be easily accomplished with our wrapper function, which takes
the same parameters as we have discussed before and returns a Seurat
object containing the results of the analysis.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sc.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/oar_by_factor.html">oar_by_factor</a></span><span class="op">(</span><span class="va">sc.data</span>, cores <span class="op">=</span> <span class="fl">1</span>, factor <span class="op">=</span> <span class="st">"group"</span>, suffix <span class="op">=</span> <span class="st">".factor"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Splitting data by specified factor..."</span></span>
<span><span class="co">#&gt; Warning in FUN(X[[i]], ...): Running process in fewer than 2 cores will considerably slow down progress</span></span>
<span><span class="co">#&gt; [1] "Extracting data..."</span></span>
<span><span class="co">#&gt; [1] "Extracting count tables"</span></span>
<span><span class="co">#&gt; [1] "Analysis started on:"</span></span>
<span><span class="co">#&gt; [1] "2025-05-23 15:41:32 UTC"</span></span>
<span><span class="co">#&gt; [1] "Identifying missing data patterns..."</span></span>
<span><span class="va">...</span></span>
<span><span class="co">#&gt; Warning in FUN(X[[i]], ...): Running process in fewer than 2 cores will considerably slow down progress</span></span>
<span><span class="co">#&gt; [1] "Extracting data..."</span></span>
<span><span class="co">#&gt; [1] "Extracting count tables"</span></span>
<span><span class="co">#&gt; [1] "Analysis started on:"</span></span>
<span><span class="co">#&gt; [1] "2025-05-23 15:41:41 UTC"</span></span>
<span><span class="co">#&gt; [1] "Identifying missing data patterns..."</span></span>
<span><span class="va">...</span></span>
<span><span class="co">#&gt; Warning in FUN(X[[i]], ...): Running process in fewer than 2 cores will considerably slow down progress</span></span>
<span><span class="co">#&gt; [1] "Extracting data..."</span></span>
<span><span class="co">#&gt; [1] "Extracting count tables"</span></span>
<span><span class="co">#&gt; [1] "Analysis started on:"</span></span>
<span><span class="co">#&gt; [1] "2025-05-23 15:41:45 UTC"</span></span>
<span><span class="co">#&gt; [1] "Identifying missing data patterns..."</span></span>
<span><span class="va">...</span></span></code></pre></div>
<p>Now let us examine the outcome of this analysis:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scatter_score_missing.html">scatter_score_missing</a></span><span class="op">(</span><span class="va">sc.data</span>, pt.size <span class="op">=</span> <span class="fl">0.5</span>, suffix <span class="op">=</span> <span class="st">".factor"</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">p4</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scatter_score_missing.html">scatter_score_missing</a></span><span class="op">(</span><span class="va">sc.data</span>, pt.size <span class="op">=</span> <span class="fl">0.5</span>, group.by <span class="op">=</span> <span class="st">"group"</span>, suffix <span class="op">=</span> <span class="st">".factor"</span><span class="op">)</span></span>
<span><span class="va">p3</span><span class="op">+</span><span class="va">p4</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figurescatter_by_condition-1.png" alt="Scatter Plots by Factors" width="100%"><p class="caption">
Scatter Plots by Factors
</p>
</div>
<p>As you can see, we are able to identify heterogeneity within each
biological condition. Moreover, we can visualize these results in a UMAP
projection of the using <code><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">Seurat::FeaturePlot()</a></code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p5</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span></span>
<span>  <span class="va">sc.data</span>, features <span class="op">=</span> <span class="st">"OARscore.factor"</span>, order <span class="op">=</span> <span class="cn">T</span>, pt.size <span class="op">=</span> <span class="fl">0.5</span>, </span>
<span>  min.cutoff <span class="op">=</span> <span class="st">"q40"</span>, max.cutoff <span class="op">=</span> <span class="st">"q90"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: The `slot` argument of `FetchData()` is deprecated as of SeuratObject 5.0.0.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Please use the `layer` argument instead.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> The deprecated feature was likely used in the <span style="color: #0000BB;">Seurat</span> package.</span></span>
<span><span class="co">#&gt;   Please report the issue at <span style="color: #0000BB; font-style: italic;">&lt;https://github.com/satijalab/seurat/issues&gt;</span>.</span></span>
<span><span class="co">#&gt; <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">generated.</span></span></span>
<span><span class="va">p5</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figureOAR_factor_featureplot-1.png" alt="Results by factor" width="100%"><p class="caption">
Results by factor
</p>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by David Sanin, Ruizhe Chen, Hannah Moore, Dmitrijs Lvovs, Paul Gueguen.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer>
</div>





  </body>
</html>
